### DISCUSSION TITLE
Cloning data

### DISCUSSION THREAD

**Discussion by SonofNun15** [AutoFixture/AutoFixture#1253] [Q&A] [answered]
Is there a way to seed a fixture with an existing model?

When testing updates, it is often useful to start with the same base model (complex type) and then change one or two properties.

Something like this:
```
var model = fixture.Create<MyModel>();
var updatedModel = fixture.Build(model).With(m => m.SomeProperty, model.SomeProperty + 1).Create();
```

Is there a way to do something like this using AutoFixture?

Thanks!

**Comment by SonofNun15** [comment]
This already exists, just found it. Use `FromSeed()`

Works like this:
```
var model = fixture.Create<MyModel>();
var updatedModel = fixture.Build<MyModel>()
  .FromSeed(m => model)
  .With(m => m.SomeProperty, model.SomeProperty + 1)
  .Create();
```

**Comment by SonofNun15** [comment]
Turns out that `FromSeed` doesn't appear to actually copy the seed object. So `With` will mutate this object, making this not ideal for setting up tests.

When I have a test for update logic, I want to take a generated object and then create a new version of this object with a few specific fields changed. I then assert that after running my update code the result is the new cloned version of the original object.

Is there a way to do this with AutoFixture? It seems like auto fixture is doing stuff very similar to this already...

**Comment by ecampidoglio** [comment, marked as answer]
AutoFixture, by design, can't guarantee that a seed value is going to be used when creating instances of a particular type. Among the built-in specimen builders, the only one that takes the seed value into account is [the one for strings](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/StringSeedRelay.cs#L7-L8).

It's up to you to define _how_ a seed value should be used for a specific type through a customization. In your case, it would involve _cloning_ the seed object, which is something you'll have to implement yourself. See [this answer](https://stackoverflow.com/a/28475970/26396) on Stack Overflow for an example on how to customize a type with a seed value.

Having said that, if I understand [your scenario](https://github.com/AutoFixture/AutoFixture/discussions/1253?converting=1#discussioncomment-573674) correctly, I think the easiest solution is to assert that the expected property has been modified correctly by using [_semantic equivalence_](https://www.rahulpnath.com/blog/semantic-comparison-improve-test-assertions/).

Based on [this](https://github.com/AutoFixture/AutoFixture/discussions/1253?converting=1#discussioncomment-573674):

> When I have a test for update logic, I want to take a generated object and then create a new version of this object with a few specific fields changed. I then assert that after running my update code the result is the new cloned version of the original object.

You could write your assertion using [SemanticComparison](https://github.com/AutoFixture/SemanticComparison)<sup>1</sup> as:

```csharp
// Arrange
var model = fixture.Create<MyModel>();

// Act
var updatedModel = sut.YourMethodThatDoesTheUpdateLogic(model);

// Assert
model
    .AsSource()
    .OfLikeness<MyModel>()
    .With(dest => dest.SomeProperty)
    .EqualsWhen((src, dest) => dest.SomeProperty == src.SomeProperty + 1)
    .ShouldEqual(updatedModel);
```

When this test fails, you'll get an error message telling you exactly which properties were different between `model` and `updatedModel`. For example:

```text
SemanticComparison.LikenessException
The provided value MyModel did not match the expected value MyModel. The following members did not match:
- SomeProperty.
```

Would that work for you?

---
1. If you're using [FluentAssertions](https://fluentassertions.com/), you can use its built-in[ object graph comparison](https://fluentassertions.com/objectgraphs/) with the `Should().BeEquivalentTo` method to do the same thing.

**Reply by SonofNun15** [reply]
Very interested in SemanticComparison... is there any documentation on this repository? The README is pretty sparse.

**Reply by SonofNun15** [reply]
Looks like your link on semantic equivalence has some interesting commentary on this. Reading through that now.

**Comment by aivascu** [comment]
@SonofNun15 just to exemplify what was said above about cloning the seed object.
Here's how your original solution would have looked like using the seeded customization.

```cs
var fixture = new Fixture();
var model = fixture.Create<MyClass>();

fixture.Customize<MyClass>(c => c
    .FromSeed(x => new MyClass(x))
    .OmitAutoProperties()
    .Do(x => x.MemberB += 1));
var model2 = fixture.Create(model);

Assert.Equal(model.MemberB + 1, model2.MemberB);
Assert.Equal(model.MemberA, model2.MemberA);
```
This is assuming your model has a copy constructor. If it does not you could probably use instead a cloning library (there are lots to choose from), but that depends on your implementation.

That said you should probably take @ecampidoglio's answer and assert your updated members.

**Reply by SonofNun15** [reply]
Great feedback, clarifies things a lot. Thank you very much!
